File: SpeedReader/RSVPEngine.swift
import Foundation
import SwiftUI
import SwiftData

@Observable
class RSVPEngine {
    var words: [String] = []
    var currentWordIndex: Int = 0
    var isPlaying: Bool = false
    
    var wpm: Double = 300 {
        didSet { if isPlaying { startTimer() } }
    }
    var wordsPerChunk: Int = 1 {
        didSet { 
            updateDisplay()
            if isPlaying { startTimer() } 
        }
    }
    var showORP: Bool = true
    var textSize: Double = 48.0
    
    var currentChunk: [String] = []
    var currentProgress: Double = 0.0
    
    private var timer: Timer?
    var currentBook: Book?
    
    func loadBook(_ book: Book) {
        self.currentBook = book
        processText(book.fullText)
        self.currentWordIndex = book.lastReadPosition
        
        if self.currentWordIndex >= self.words.count {
            self.currentWordIndex = 0
        }
        
        self.isPlaying = false
        stopTimer()
        updateDisplay()
    }
    
    private func processText(_ text: String) {
        // Clean punctuation: remove commas, periods, exclamation marks
        let punctuationToRemove = CharacterSet(charactersIn: ".,/#!$%^&*;:{}=\\-_`~()\"'“”")
        let cleanedText = text.components(separatedBy: punctuationToRemove).joined(separator: " ")
        
        // Split into words
        self.words = cleanedText.components(separatedBy: .whitespacesAndNewlines).filter { !$0.isEmpty }
    }
    
    func jumpTo(index: Int) {
        guard !words.isEmpty else { return }
        currentWordIndex = max(0, min(index, words.count - 1))
        updateDisplay()
        saveProgress()
    }
    
    func jumpByPercentage(_ percentage: Double) {
        let index = Int(Double(words.count) * percentage)
        jumpTo(index: index)
    }
    
    func jumpByVirtualPage(forward: Bool) {
        let pageSize = 500
        let delta = forward ? pageSize : -pageSize
        jumpTo(index: currentWordIndex + delta)
    }
    
    func toggleBookmark() {
        guard let book = currentBook else { return }
        if let existing = book.bookmarks.firstIndex(of: currentWordIndex) {
            book.bookmarks.remove(at: existing)
        } else {
            book.bookmarks.append(currentWordIndex)
            book.bookmarks.sort()
        }
    }
    
    func playPause() {
        guard !words.isEmpty else { return }
        isPlaying.toggle()
        if isPlaying {
            startTimer()
        } else {
            stopTimer()
        }
    }
    
    func reset() {
        stopTimer()
        currentWordIndex = 0
        updateDisplay()
        saveProgress()
    }
    
    private func startTimer() {
        stopTimer()
        let intervalPerWord = 60.0 / wpm
        let timerInterval = intervalPerWord * Double(wordsPerChunk)
        
        timer = Timer.scheduledTimer(withTimeInterval: timerInterval, repeats: true) { [weak self] _ in
            self?.advance()
        }
    }
    
    private func stopTimer() {
        timer?.invalidate()
        timer = nil
        isPlaying = false
    }
    
    private func advance() {
        if currentWordIndex + wordsPerChunk >= words.count {
            currentWordIndex = max(0, words.count - 1)
            stopTimer()
        } else {
            currentWordIndex += wordsPerChunk
        }
        updateDisplay()
        saveProgress()
    }
    
    func updateDisplay() {
        guard !words.isEmpty else {
            currentChunk = []
            currentProgress = 0
            return
        }
        
        let endIndex = min(currentWordIndex + wordsPerChunk, words.count)
        currentChunk = Array(words[currentWordIndex..<endIndex])
        currentProgress = Double(currentWordIndex) / Double(max(1, words.count - 1))
    }
    
    func saveProgress() {
        if let book = currentBook {
            book.lastReadPosition = currentWordIndex
        }
    }
}

File: SpeedReaderAndroid/RSVPEngine.kt
package com.example.speedreader

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlin.math.max
import kotlin.math.min

class RSVPEngine : ViewModel() {

    var words = emptyList<String>()
    var currentBook: Book? = null

    private val _currentChunk = MutableStateFlow<List<String>>(emptyList())
    val currentChunk: StateFlow<List<String>> = _currentChunk.asStateFlow()

    private val _isPlaying = MutableStateFlow(false)
    val isPlaying: StateFlow<Boolean> = _isPlaying.asStateFlow()

    private val _currentProgress = MutableStateFlow(0f)
    val currentProgress: StateFlow<Float> = _currentProgress.asStateFlow()

    private val _currentWordIndex = MutableStateFlow(0)
    val currentWordIndex: StateFlow<Int> = _currentWordIndex.asStateFlow()

    // Settings
    var wpm = MutableStateFlow(300f)
    var wordsPerChunk = MutableStateFlow(1)
    var textSize = MutableStateFlow(56f)
    var showORP = MutableStateFlow(true)
    var focalColor = MutableStateFlow(0xFFEAA023)
    var textColor = MutableStateFlow(0xFFFFFFFF)
    var textBgColor = MutableStateFlow(0x00000000)
    var isDarkTheme = MutableStateFlow(true)

    private var timerJob: Job? = null

    fun loadBook(book: Book) {
        currentBook = book
        processText(book.fullText)
        _currentWordIndex.value = book.lastReadPosition
        
        if (_currentWordIndex.value >= words.size) {
            _currentWordIndex.value = 0
        }

        stopTimer()
        updateDisplay()
    }

    private fun processText(text: String) {
        val cleanRegex = Regex("[.,/#!$%^&*;:{}=\\\\-_`~()\"'“”]")
        val cleanedText = text.replace(cleanRegex, " ")
        words = cleanedText.split(Regex("\\s+")).filter { it.isNotEmpty() }
    }

    fun jumpTo(index: Int) {
        _currentWordIndex.value = max(0, min(index, words.size - 1))
        updateDisplay()
        saveProgress()
    }

    fun jumpByVirtualPage(forward: Boolean) {
        val pageSize = 500
        val delta = if (forward) pageSize else -pageSize
        jumpTo(_currentWordIndex.value + delta)
    }

    fun toggleBookmark() {
        val book = currentBook ?: return
        val current = _currentWordIndex.value
        val newList = book.bookmarks.toMutableList()
        if (newList.contains(current)) {
            newList.remove(current)
        } else {
            newList.add(current)
            newList.sort()
        }
        book.bookmarks = newList
    }

    fun togglePlayPause() {
        if (words.isEmpty()) return
        if (_isPlaying.value) stopTimer() else startTimer()
    }

    fun reset() {
        stopTimer()
        _currentWordIndex.value = 0
        updateDisplay()
        saveProgress()
    }

    fun startTimer() {
        if (words.isEmpty()) return
        _isPlaying.value = true
        timerJob?.cancel()
        timerJob = viewModelScope.launch {
            while (isActive && _isPlaying.value) {
                val interval = 60000L / wpm.value.toLong() * wordsPerChunk.value
                delay(interval)
                advance()
            }
        }
    }

    fun stopTimer() {
        _isPlaying.value = false
        timerJob?.cancel()
        timerJob = null
    }

    private fun advance() {
        if (_currentWordIndex.value + wordsPerChunk.value >= words.size) {
            _currentWordIndex.value = max(0, words.size - 1)
            stopTimer()
        } else {
            _currentWordIndex.value += wordsPerChunk.value
        }
        updateDisplay()
        saveProgress()
    }

    fun updateDisplay() {
        if (words.isEmpty()) {
            _currentChunk.value = emptyList()
            _currentProgress.value = 0f
            return
        }
        val startIndex = _currentWordIndex.value
        val endIndex = min(startIndex + wordsPerChunk.value, words.size)
        _currentChunk.value = words.subList(startIndex, endIndex)
        _currentProgress.value = startIndex.toFloat() / max(1, words.size - 1).toFloat()
    }

    fun saveProgress() {
        currentBook?.lastReadPosition = _currentWordIndex.value
        // In a real app, you'd trigger a DAO update here.
    }
}
